### PR TITLE
Rework api/user/repos for pagination

### LINKED ISSUE #11800: [Bug] [API] make pagination optional for GET /api/user/repos
come up with #9452 in v1.12
ref https://github.com/go-gitea/gitea/pull/9452#issuecomment-640319866

### PR DESCRIPTION
**Pull request by CirnoT**
1. Add count to `GetUserRepositories` so that pagination can be supported for `/user/{username}/repos`
2. Rework `ListMyRepos` to use `models.SearchRepository`

---

`ListMyRepos` was an odd one. It first fetched all user repositories and then tried to supplement them with accessible map. The end result was that:
- Limit for pagination did not work because accessible repos would always be appended
- The amount of pages was incorrect if one were to calculate it
- When paginating, all accessible repos would be shown on every page

Hopefully it should now work properly. Fixes #11800 and does not require any change on Drone-side as it can properly interpret and act on `Link` header which we now set.

### CODE REVIEW

**Comment by CirnoT** [comment]
Needs to be backported to v1.12 as to not break Drone integration.

File: routers/api/v1/user/repo.go [hunk 1]
```diff
@@ -36,6 +39,9 @@ func listUserRepos(ctx *context.APIContext, u *models.User, private bool) {
 			apiRepos = append(apiRepos, repos[i].APIFormat(access))
 		}
 	}
+
+	ctx.SetLinkHeader(int(count), opts.PageSize)
+	ctx.Header().Set("X-Total-Count", fmt.Sprintf("%d", count))
```
Comment (lafriks) [inline routers/api/v1/user/repo.go]:
```suggestion
	ctx.Header().Set("X-Total-Count", strconv.FormatInt(count, 10))
```

File: routers/api/v1/user/repo.go [hunk 2]
```diff
@@ -92,31 +98,37 @@ func ListMyRepos(ctx *context.APIContext) {
...
+		accessMode, err := models.AccessLevel(ctx.User, repo)
+		if err != nil {
+			ctx.Error(http.StatusInternalServerError, "AccessLevel", err)
+		}
+		results[i] = repo.APIFormat(accessMode)
 	}
-	ctx.JSON(http.StatusOK, &apiRepos)
+
+	ctx.SetLinkHeader(int(count), opts.ListOptions.PageSize)
+	ctx.Header().Set("X-Total-Count", fmt.Sprintf("%d", count))
```
Comment (lafriks) [inline routers/api/v1/user/repo.go]:
```suggestion
	ctx.Header().Set("X-Total-Count", strconv.FormatInt(count, 10))
```

File: routers/api/v1/user/repo.go [hunk 3]
```diff
@@ -92,31 +98,37 @@ func ListMyRepos(ctx *context.APIContext) {
...
-	for i := range ownRepos {
-		apiRepos[i] = ownRepos[i].APIFormat(models.AccessModeOwner)
-	}
-	i := len(ownRepos)
-	for repo, access := range accessibleReposMap {
-		apiRepos[i] = repo.APIFormat(access)
-		i++
+	results := make([]*api.Repository, len(repos))
+	for i, repo := range repos {
+		if err = repo.GetOwner(); err != nil {
```
Comment (lafriks) [inline routers/api/v1/user/repo.go:118]: Would be good if this could be done on all list at once

Comment (CirnoT) [reply routers/api/v1/user/repo.go:118]: Hm? FYI this is direct copy from search API

Comment (lafriks) [reply routers/api/v1/user/repo.go:118]: This was more of the suggestion not and probably is out of scope for this PR

**Comment by lafriks** [comment]
lint error about unused "fmt" import

**Review by lafriks — APPROVED** [review]

**Review by zeripath — APPROVED** [review]
OK this has waited long enough for dronistas to pipe up and say this further breaks things. LGTM.

**Comment by zeripath** [comment]
make lg-tm work

**Comment by zeripath** [comment]
@CirnoT please send backport

**Pull request merged**